hibernate-core/src/main/java/org/hibernate/proxy/pojo/javassist/JavassistProxyFactory.java

```diff
@@ -93,7 +93,21 @@ public void postInstantiate(
 		javassist.util.proxy.ProxyFactory factory = new javassist.util.proxy.ProxyFactory() {
 			@Override
 			protected ClassLoader getClassLoader() {
-				return persistentClass.getClassLoader();
+				// This classloader does not take into account class space consistency
+				// i.e. if the persistenceClass can see the same interface types via
+				// a different ClassLoader hierarchy. In that case we could try some 
+				// risky magic with loadClass, but using findClass is much safer.
+				return new ClassLoader(persistentClass.getClassLoader()) {
```

Comment:
> @brmeyer And I am a little shaky on what exactly (Class loading wise) Javassist should be given access to here :)
I need to understand that better to give you a better answer.

There are three sets of types that need to be unified by the Proxy ClassLoader. 

1. The first set is obviously the persistent class type, and things that it implicitly references. This can be handled by either the persistent class class loader (i.e. what's being used now), or by the persistence unit class loader.

2. The second set of classes that need to be loadable are the proxy interfaces being added by Hibernate. There only appears to be one interface `org.hibernate.proxy.HibernateProxy`, which directly references (`org.hibernate.proxy.LazyInitializer`). This list may change over time if this area changes.

3. The third set of classes that need to be loaded are Javassist classes, these include things like `javassist.util.proxy.Proxy` and `javassist.util.proxy.ProxyObject`

> I think creating the new ClassLoader each and every call is going to be bad here. Javassist uses this as a cache key; we might get lucky there in that we really do not need the "ProxyFactory caching" based on how we use Javassist. I am missing the logic why a new ClassLoader each time is being proposed.

There isn't a particular reason to use a new ClassLoader each time, and I agree that the current PR may have negative side-effects if Javassist is caching by ClassLoader. My aim was to find the least invasive solution and to avoid breaking anything else - if there's a better way to achieve the same thing then I'm all for it.